docs: release template: update based on 1.28 learnings (#12356)
filecoin-project/lotus: documentation/misc/RELEASE_ISSUE_TEMPLATE.md

> Release Issue Template

# Lotus Node|Miner X.Y.Z Release

[//]: # (Below are non-visible steps intended for the issue creator)
[//]: # (Start an issue with title "[WIP] Lotus Node|Miner vX.Y.Z" and adjust the title for whether it's a Node or Miner release.)
[//]: # (Copy in the content of https://github.com/filecoin-project/lotus/blob/master/documentation/misc/RELEASE_ISSUE_TEMPLATE.md)
[//]: # (Find/Replace "X.Y.Z" with the actual values.)
[//]: # (Copy/paste the "Release Checklist > RCX" section to "Release Checklist > Stable \(non-RC\) Release" and apply the "diff" called out there.)
[//]: # (Find/Replace "RCX" with "RC1".)
[//]: # (If this isn't a release tied to a network upgrade, remove all items with "\(network upgrade\)")
[//]: # (Adjust the "Meta" section values.)
[//]: # (Apply the `tpm` label to the issue)
[//]: # (Create the issue)
[//]: # (Pin the issue on GitHub)

## 😶‍🌫 Meta
* Scope: Node|Miner

> * Is this linked with a netwok upgrade, and thus mandatory? Yes|No

Reviewer: ```suggestion
* Is this linked with a network upgrade, and thus mandatory? Yes|No
```

## 😶‍🌫 Meta
* Scope: Node|Miner
* Is this linked with a netwok upgrade, and thus mandatory? Yes|No
* Related network upgrade version: n/a|nvXX

## 🚢 Estimated shipping date

<Date this release will ship on if everything goes to plan (week beginning...)>

## ✅ Release Checklist

**Note for whoever is owning the release:** please capture notes as comments in this issue for anything you noticed that could be improved for future releases.  There is a *Post Release* step below for incorporating changes back into the [RELEASE_ISSUE_TEMPLATE](https://github.com/filecoin-project/lotus/blob/master/documentation/misc/RELEASE_ISSUE_TEMPLATE.md), and this is easier done by collecting notes from along the way rather than just thinking about it at the end.
[//]: # (If/when we know an exact date, remove the "week of".)
[//]: # (If a date week is an estimate, annotate with "estimate".)

> - [ ] Fork a new branch (`release/vX.Y.Z` or `release/miner/vX.Y.Z`) from `master` and make any further release related changes to this branch. If any "non-trivial" changes get added to the release, uncheck all the checkboxes and return to this stage.

Author: > If any "non-trivial" changes get added to the release, uncheck all the checkboxes and return to this stage.

Instead of an "uncheck" flow, I am going with a "copy a new section" flow so it's clear what RC we're on and what has been done.

Reviewer: I like this. Additionally it makes it clearer if release engineer needs to hand over the release duties in the middle of a RC-train

> > 3. They are presented here for quick reference, but backporting is tracked in each `Release Checklist`.

Author: I went back and forth about tracking these here vs. adding them to each RC section below in Release Chcklist.  I figured it was useful to see what changes we are staging for which RC in a consolidated place like we did https://github.com/filecoin-project/lotus/issues/12109 .  I am also fine to inline it.

Author: Per 2024-08-09 verbal, we're going to go with this approach for now and can adjust as we do a 1.29.0 release.

[//]: # (Copy/paste this for each RC, and increment "X")
### RCX
- [ ] To Be Added

### Stable (non-RC)
- [ ] To Be Added

## ✅ Release Checklist

### Before RC1

> - [ ] (network upgrade) Make sure all [Lotus dependecies are updated to the correct versions for the network upgrade](https://github.com/filecoin-project/lotus/blob/master/documentation/misc/Update_Dependencies_Lotus.md)

Reviewer: ```suggestion
- [ ] (network upgrade) Make sure all [Lotus dependencies are updated to the correct versions for the network upgrade](https://github.com/filecoin-project/lotus/blob/master/documentation/misc/Update_Dependencies_Lotus.md)
```

   - Link to PR:
- [ ] Open PR against [RELEASE_ISSUE_TEMPLATE.md](https://github.com/filecoin-project/lotus/blob/master/documentation/misc/RELEASE_ISSUE_TEMPLATE.md) with title `docs(vX.Y.Z): release template improvements` for improving future releases.
   - Link to PR:  
   - This will get merged in a `Post Release` step, but improvements are better done by collecting notes along the way rather than just thinking about it at the end.
- [ ] Fork a new branch (`release/vX.Y.Z` or `release/miner/vX.Y.Z`) from `master` and make any further release-related changes to this branch.

> - [ ] In the `master` branch, bump the version in `build/version.go`  to `vX.Y.(Z+1)-dev`. Run `make gen && make docsgen-cli` before committing changes.

Reviewer: I think we can make it clearer here that the version string in:
- https://github.com/filecoin-project/lotus/blob/4b75fa86b2567409b815ac4627256bd275e65fa3/build/version.go#L10 should be updated if shipping a node release.
- https://github.com/filecoin-project/lotus/blob/4b75fa86b2567409b815ac4627256bd275e65fa3/build/version.go#L21 should be updated if shipping a miner release.

Reviewer: Actually, since this is the `master` we should be fine with saying something like "bump the version for the type of release you are shipping node|miner"

Author: Good point - addressed in https://github.com/filecoin-project/lotus/pull/12356/commits/3484d63c1eb1a781d917b8f442bb29136a031fbc

   - Link to PR: 

### RCs

> [//]: # (Copy/paste this "RCX" section for each RC, and increment "X")

Author: ```suggestion
[//]: # (Copy/paste this "RCX" section for each additional RC, and increment "X")
```

#### RCX
> [!IMPORTANT]
> These PRs should be done in and target the `release/vX.Y.Z` or `release/miner/vX.Y.Z` branch.

> **Backport PR**

Author: I assume there won't be a "Backport PR" for RC1 though since we won't start RC1 until items land.    Hmm, I'm not sure the best way to communicate that creating a PR here is for RC2 onwards.  For RC1, we still need to make sure the tracked items still land though.  (I'll give this fresh thought 2024-08-08.)

Author: Best approach I could come up with was to add comments about how RC1 is a bit special here: https://github.com/filecoin-project/lotus/pull/12356/commits/be71e636bd3f5ec4a7726b41fa291212a1c0db63

- [ ] All explicitly tracked items from `Dependencies for releases` have landed
- [ ] Backported [everything with the "backport" label](https://github.com/filecoin-project/lotus/issues?q=label%3Arelease%2Fbackport+) 
- [ ] Removed the "backport" label from all backported PRs (no ["backport" issues](https://github.com/filecoin-project/lotus/issues?q=label%3Arelease%2Fbackport+))

> - [ ] Create a PR with title `build(vX.Y.Z): backport changes for vX.Y.Z-rcX`

Author: PR title here depends on what is decided in https://github.com/filecoin-project/lotus/pull/12340 .  (I went with conventional commit syntax for now.)

> - [ ] Create a PR with title `build(vX.Y.Z): backport changes for vX.Y.Z-rcX`

Author: This is where we can add extra notes in future about how we're backporting (e.g., cherrypicking, merge commits).

   - Link to PR: 
- [ ] Merge PR 

**Release PR**
- [ ] Update the version string in `build/version.go` to one ending with '-rcX'. 
    - Ensure to update the appropriate version string based on whether you are creating a node release (`NodeBuildVersion`), a miner release (`MinerBuildVersion`), or both.
- [ ] Run `make gen && make docsgen-cli` to generate documentation

> - [ ] Changelog prep

Author: Is it right that this should be happening in `release/vX.Y.Z` rather than in `master` and then cherry picked.

Reviewer: It is correct in the sense that this is what we have done previously. I do not have any particular stance on whether we should keep it the current way or change to land in master and then cherry-pick."

Author: Hmm, thinking on this more...

If we do the changelog updeates in release/vX.Y.Z then if someone looks at the the changelog in the master branch, they won't actually see the latest changelog for an in process release.   That's not ideal, but I assume we can live with it since the GitHub releases themsleves have the curated/accurate changelog text in the release notes.

If we do the changes to the changelog in master, then we need backport the changelog change when we backport other changes.  That potentially gets a bit messy if there are other changelog entries that have been added in the interim for the "unreleased version".

I don't have a strong thought here so maybe we just leave as-is currently and assess this next time through.

   - [ ] Generate changelog using the script at `scripts/mkreleaselog`
   - [ ] Add contents of generated text to `CHANGELOG.md`
   - [ ] Editorializing (breaking changes, new feature callouts, FIPs, actor bundles, new feature callouts)
   - [ ] (network upgrade) Specify whether the Calibration or Mainnet upgrade epoch has been specified or not yet.
      - Example where these weren't specified yet: [PR #12169](https://github.com/filecoin-project/lotus/pull/12169)
- [ ] Create a PR with title `build(vX.Y.Z): release vX.Y.Z-rcX`
   - Link to PR: 
   - Opening a PR will trigger a CI run that will build assets, create a draft GitHub release, and attach the assets.
- [ ] Merge the PR

>    - Opening the PR will trigger a CI run that will build assets, attach the assets to the GitHub release, public the GitHub release, and create the corresponding git tag.

Reviewer: ```suggestion
   - Opening the PR will trigger a CI run that will build assets, attach the assets to the GitHub release, publish the GitHub release, and create the corresponding git tag.
```

 - [ ] Update `🚢 Estimated shipping date` table
 - [ ] Comment on this issue announcing the RC
    - Link to issue comment:   

### Testing
> [!NOTE]
> Link to any special steps for testing releases beyond ensuring CI is green.  Steps can be inlined here or tracked elsewhere.

### Stable (non-RC) Release

> [//]: # (This "note" with the "diff" to apply to the RC is here to avoid the duplication in the template itself.)

Author: ```suggestion
[//]: # (This "NOTE" below with the "diff" to apply to the "RC copy/pasted content" is here to avoid the duplication in the template itself.)
```

> [//]: # (This is done as a visible NOTE rather than a comment to make sure it's clear what needs to be added to this section.)

Author: Feedback welcome if you are ok with this approach.  As I was working on this PR, I was getting tired of having duplication in this "Stable" section with what was in the "RCX" section.

Reviewer: I´m okay with this approach. If we find out that this is confusing to someone new to the flow, we can always adjust (add duplication) later.

> [//]: # (These comment ^^^ can be removed once the NOTE steps are completed.)

Author: ```suggestion
[//]: # (These comments ^^^ can be removed once the NOTE steps below are completed.)
```

> [!NOTE]
> Copy/paste in the `RCX` section above and then make these changes:
> 1. Change the version string text:
> 
> Update the version string in `build/version.go` to one **NOT** ending with '-rcX'
>
> 2. Under "Changelog prep", add
> 
> (network upgrade) Ensure the Mainnet upgrade epoch is specified.
> 
> 3. Remove this `[!Note]` and the related invisible comments.

### Post-Release

> - [ ] Open a pull request against `master` with a merge of the `release/vX.Y.Z` branch. 

Author: Note that I'm currently skipping anything with `releases`.  I figured we can adjust once this is evaluated more as part of https://github.com/filecoin-project/lotus/issues/12072

> - [ ] Open a pull request against `master` with a merge of the `release/vX.Y.Z` branch. 

Author: I'd ideally like to list the specific command that should be run here so it's clear whether we're cherry picking, merge commit.  This isn't any less detail than what we had before though, so it can be improved after once we are clear and documentwhat commit types we use and when.

[//]: # (For RC1 there likely isn't any backporting to do and thus no PR which reduces the steps.)
[//]: # (We do need all these steps for RC2 onwards though.)
[//]: # (If steps are removed for the RC1 checklist, they need to be preserved for future RCs/stable.)
[//]: # (For RC1 we still need to make sure the tracked items land though.)
- [ ] All explicitly tracked items from `Dependencies for releases` have landed
- [ ] Backported [everything with the "backport" label](https://github.com/filecoin-project/lotus/issues?q=label%3Arelease%2Fbackport+) 
- [ ] Removed the "backport" label from all backported PRs (no ["backport" issues](https://github.com/filecoin-project/lotus/issues?q=label%3Arelease%2Fbackport+))
- [ ] Create a PR with title `build(vX.Y.Z): backport changes for vX.Y.Z-rcX`
   - Link to PR: 
- [ ] Merge PR 

**Release PR**

- [ ] Update the version string(s) in `build/version.go` to one ending with '-rcX'. 
    - Ensure to update the appropriate version string based on whether you are creating a node release (`NodeBuildVersion`), a miner release (`MinerBuildVersion`), or both.
- [ ] Run `make gen && make docsgen-cli` to generate documentation
- [ ] Changelog prep

>    - [ ] Generate changelog using the script at `scripts/mkreleaselog`

Reviewer: I think we can remove this step since we now get the changelog automatically generated for the release/tag when we open a PR that changes the version string: https://github.com/filecoin-project/lotus/pull/12096.

One problem that arises is how we populate this back into the CHANGELOG.md file in Lotus, though.

Author: Oh really - that doesn't seem good/right.  I assume we want the text for the release notes to be under code review, which means its source of truth should be changelog.  I'll look into this more as thought we hit this recently with one of the other repos as well.

Author: Oh doh - your comment is about just this line of `scripts/mkreleaselog`.  My bad.  I misread this in how it was coming through in the GitHub conversation view (not files view).  I agree this isn't needed anymore: https://github.com/filecoin-project/lotus/blob/master/.github/workflows/release.yml#L256 .  I'll remove it now.

Author: commit with adjust changelog prep steps: https://github.com/filecoin-project/lotus/pull/12356/commits/e3805466b376aa56d006cf471adf4c1d131fc384

I also added commands there that can run for looking at the history to help identify if anything missing in the changelog.  These were just ideas I would do off the top of my head but maybe you had other commands you'd run?
